File: psd_max_channels.m
%从功率谱密度（psd）中挑选出5个频率段下每个时间段能量最大的通道
theta = zeros(2, 24, 80, 2, 20);
alpha = zeros(2, 24, 80, 2, 20);
lowbeta = zeros(2, 24, 80, 2, 20);
highbeta = zeros(2, 24, 80, 2, 20);
lowgamma = zeros(2, 24, 80, 2, 20);
highgamma = zeros(2, 24, 80, 2, 20);
%condition 1
for i=1:20
    load(['stft_VP', sprintf('%02d', i), '_psd_con1.mat']);
    for j=1:80
        for k=1:24
            psdInTheta = reshape(psd(5:8,k,:,j),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,j,1,i) = bandMax;
            theta(2,k,j,1,i) = maxCh;
            psdInAlpha = reshape(psd(9:13,k,:,j),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,j,1,i) = bandMax;
            alpha(2,k,j,1,i) = maxCh;
            psdInLowbeata = reshape(psd(15:21,k,:,j),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,j,1,i) = bandMax;
            lowbeta(2,k,j,1,i) = maxCh;
            psdInHighbeata = reshape(psd(21:31,k,:,j),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,j,1,i) = bandMax;
            highbeta(2,k,j,1,i) = maxCh;
            psdInLowgamma = reshape(psd(31:61,k,:,j),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,j,1,i) = bandMax;
            lowgamma(2,k,j,1,i) = maxCh;
            psdInHighgamma = reshape(psd(61:101,k,:,j),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,j,1,i) = bandMax;
            highgamma(2,k,j,1,i) = maxCh;
        end
    end
end

%condition 2
for i=1:20
    load(['stft_VP', sprintf('%02d', i), '_psd_con2.mat']);
    for j=1:80
        for k=1:24
            psdInTheta = reshape(psd(5:8,k,:,j),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,j,2,i) = bandMax;
            theta(2,k,j,2,i) = maxCh;
            psdInAlpha = reshape(psd(9:13,k,:,j),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,j,2,i) = bandMax;
            alpha(2,k,j,2,i) = maxCh;
            psdInLowbeata = reshape(psd(15:21,k,:,j),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,j,2,i) = bandMax;
            lowbeta(2,k,j,2,i) = maxCh;
            psdInHighbeata = reshape(psd(21:31,k,:,j),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,j,2,i) = bandMax;
            highbeta(2,k,j,2,i) = maxCh;
            psdInLowgamma = reshape(psd(31:61,k,:,j),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,j,2,i) = bandMax;
            lowgamma(2,k,j,2,i) = maxCh;
            psdInHighgamma = reshape(psd(61:101,k,:,j),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,j,2,i) = bandMax;
            highgamma(2,k,j,2,i) = maxCh;
        end
    end
end
save('Channels_max_psd.mat','theta','alpha','lowbeta','highbeta','lowgamma','highgamma')

File: python_preprocessing/psd_max_channels.py
import h5py
import hdf5storage
import numpy as np

theta = np.zeros((2, 24, 80, 2, 20))
alpha = np.zeros((2, 24, 80, 2, 20))
lowbeta = np.zeros((2, 24, 80, 2, 20))
highbeta = np.zeros((2, 24, 80, 2, 20))
lowgamma = np.zeros((2, 24, 80, 2, 20))
highgamma = np.zeros((2, 24, 80, 2, 20))

for i in range(20):
    f=h5py.File('..\stft_VP{:0>2d}_psd_con1.mat'.format(i+1),'r')
    psd=f['psd']
    for j in range(80):
        for k in range(24):
            psdInTheta=np.reshape(psd[j,:,k,4:8],(128,4))
            eachMax=np.amax(psdInTheta,0)
            eachCh=np.argmax(psdInTheta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            theta[0,k,j,0,i]=bandMax
            theta[1,k,j,0,i]=maxCh
            psdInAlpha=np.reshape(psd[j,:,k,8:13],(128,5))
            eachMax=np.amax(psdInAlpha,0)
            eachCh=np.argmax(psdInAlpha,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            alpha[0,k,j,0,i]=bandMax
            alpha[1,k,j,0,i]=maxCh
            psdInLowbeta=np.reshape(psd[j,:,k,14:21],(128,7))
            eachMax=np.amax(psdInLowbeta,0)
            eachCh=np.argmax(psdInLowbeta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            lowbeta[0,k,j,0,i]=bandMax
            lowbeta[1,k,j,0,i]=maxCh
            psdInHighbeta=np.reshape(psd[j,:,k,20:31],(128,11))
            eachMax=np.amax(psdInHighbeta,0)
            eachCh=np.argmax(psdInHighbeta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            highbeta[0,k,j,0,i]=bandMax
            highbeta[1,k,j,0,i]=maxCh
            psdInLowgamma=np.reshape(psd[j,:,k,30:61],(128,31))
            eachMax=np.amax(psdInLowgamma,0)
            eachCh=np.argmax(psdInLowgamma,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            lowgamma[0,k,j,0,i]=bandMax
            lowgamma[1,k,j,0,i]=maxCh
            psdInHighgamma=np.reshape(psd[j,:,k,60:101],(128,41))
            eachMax=np.amax(psdInHighgamma,0)
            eachCh=np.argmax(psdInHighgamma,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            highgamma[0,k,j,0,i]=bandMax
            highgamma[1,k,j,0,i]=maxCh

for i in range(20):
    f=h5py.File('..\stft_VP{:0>2d}_psd_con2.mat'.format(i+1),'r')
    psd=f['psd']
    for j in range(80):
        for k in range(24):
            psdInTheta=np.reshape(psd[j,:,k,4:8],(128,4))
            eachMax=np.amax(psdInTheta,0)
            eachCh=np.argmax(psdInTheta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            theta[0,k,j,1,i]=bandMax
            theta[1,k,j,1,i]=maxCh
            psdInAlpha=np.reshape(psd[j,:,k,8:13],(128,5))
            eachMax=np.amax(psdInAlpha,0)
            eachCh=np.argmax(psdInAlpha,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            alpha[0,k,j,1,i]=bandMax
            alpha[1,k,j,1,i]=maxCh
            psdInLowbeta=np.reshape(psd[j,:,k,14:21],(128,7))
            eachMax=np.amax(psdInLowbeta,0)
            eachCh=np.argmax(psdInLowbeta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            lowbeta[0,k,j,1,i]=bandMax
            lowbeta[1,k,j,1,i]=maxCh
            psdInHighbeta=np.reshape(psd[j,:,k,20:31],(128,11))
            eachMax=np.amax(psdInHighbeta,0)
            eachCh=np.argmax(psdInHighbeta,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            highbeta[0,k,j,1,i]=bandMax
            highbeta[1,k,j,1,i]=maxCh
            psdInLowgamma=np.reshape(psd[j,:,k,30:61],(128,31))
            eachMax=np.amax(psdInLowgamma,0)
            eachCh=np.argmax(psdInLowgamma,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            lowgamma[0,k,j,1,i]=bandMax
            lowgamma[1,k,j,1,i]=maxCh
            psdInHighgamma=np.reshape(psd[j,:,k,60:101],(128,41))
            eachMax=np.amax(psdInHighgamma,0)
            eachCh=np.argmax(psdInHighgamma,0)
            bandMax=np.amax(eachMax)
            index=np.argmax(eachMax)
            maxCh=eachCh[index]
            highgamma[0,k,j,1,i]=bandMax
            highgamma[1,k,j,1,i]=maxCh
            
hdf5storage.savemat('..\Channels_max_psd.mat', {'theta':theta, 'alpha':alpha, 'lowbeta':lowbeta, 'highbeta':highbeta, 'lowgamma':lowgamma,'highgamma':highgamma})